### PR TITLE
Add execute permission to all files recursively under /install/postscripts

### PR DESCRIPTION
**Pull request by peterwywong**
The PR is to fix issue 6579.

### CODE REVIEW

**Comment by peterwywong** [comment]
Fri Apr  3 21:23:38 EDT 2020 [info]: xcat.deployment.postscript: postscript syncfiles return with 0
Fri Apr  3 21:23:38 EDT 2020 [info]: xcat.deployment.postscript: Running postscript: dir1/dir2/dir3/foo.bar
/xcatpost/mypostscript: line 41: ./dir1/dir2/dir3/foo.bar: Permission denied
Fri Apr  3 21:23:38 EDT 2020 [info]: xcat.deployment.postscript: postscript dir1/dir2/dir3/foo.bar return with 126
Fri Apr  3 21:23:38 EDT 2020 [info]: xcat.deployment.postscript: Running postscript: dir1/dir2/dir3/foo.bar
/xcatpost/mypostscript: line 41: ./dir1/dir2/dir3/foo.bar: Permission denied

**Comment by peterwywong** [comment]
Here is xcat.log for RHEL 8.1 diskfull installation with PR 6669 running "chmod -R +x /xcatpost ..." in post.xcat.ng.

Running Kickstart Post-installation script...
Fri Apr 10 10:46:24 EDT 2020 [info]: xcat.deployment: Executing post.xcat to prepare for firstbooting ...
Fri Apr 10 10:46:51 EDT 2020 [info]: xcat.deployment: trying to download postscripts from 10.3.11.15...
Fri Apr 10 10:46:52 EDT 2020 [info]: xcat.deployment: postscripts downloaded successfully
Fri Apr 10 10:46:52 EDT 2020 [info]: xcat.deployment: trying to get mypostscript from 10.3.11.15...
Fri Apr 10 10:46:52 EDT 2020 [info]: xcat.deployment.postscript: Running postscript: syslog
grep: /etc/rsyslog.d/remote.conf: No such file or directory
grep: /etc/rsyslog.d/remote.conf: No such file or directory
Fri Apr 10 10:46:52 EDT 2020 [info]: xcat.deployment.postscript: postscript syslog return with 0
Fri Apr 10 10:46:52 EDT 2020 [info]: xcat.deployment.postscript: Running postscript: remoteshellUnable to load host key: /etc/ssh/ssh_host_ed25519_key^M
Fri Apr 10 10:46:54 EDT 2020 [info]: xcat.deployment.postscript: postscript remoteshell return with 0
Fri Apr 10 10:46:54 EDT 2020 [info]: xcat.deployment.postscript: Running postscript: syncfiles
  Did not sync any files.
Fri Apr 10 10:46:54 EDT 2020 [info]: xcat.deployment.postscript: postscript syncfiles return with 0
Fri Apr 10 10:46:54 EDT 2020 [info]: xcat.deployment.postscript: Running postscript: dir1/dir2/dir3/foo.bar
"The number is NOT 1"
Fri Apr 10 10:46:54 EDT 2020 [info]: xcat.deployment.postscript: postscript dir1/dir2/dir3/foo.bar return with 1
Fri Apr 10 10:46:54 EDT 2020 [info]: xcat.deployment.postscript: Running postscript: dir1/dir2/dir3/foo.bar
"The number is ONE"
Fri Apr 10 10:46:54 EDT 2020 [info]: xcat.deployment.postscript: postscript dir1/dir2/dir3/foo.bar return with 0
Fri Apr 10 10:46:54 EDT 2020 [info]: xcat.deployment.postscript: Running postscript: dir1/dir2/dir3/foo.bar
"The number is NOT 1"
Fri Apr 10 10:46:54 EDT 2020 [info]: xcat.deployment.postscript: postscript dir1/dir2/dir3/foo.bar return with 1
Fri Apr 10 10:46:54 EDT 2020 [info]: xcat.deployment: finished firstboot preparation, sending request to c910f03c11k15:3002 for changing status...
Fri Apr 10 10:47:57 EDT 2020 [info]: xcat.deployment: Running /xcatpost/mypostscript.post
Fri Apr 10 10:47:57 EDT 2020 [info]: xcat.deployment.postbootscript: Running postbootscript: otherpkgs
Fri Apr 10 10:47:57 EDT 2020 [info]: xcat.deployment.postbootscript: postbootscript otherpkgs return with 0
Fri Apr 10 10:47:57 EDT 2020 [error]: xcat.deployment.postbootscript: provision completed with error.(c910f03c11k17)
Fri Apr 10 10:47:57 EDT 2020 [info]: xcat.deployment: /xcatpost/mypostscript.post return
Fri Apr 10 10:47:57 EDT 2020 [info]: xcat.deployment: =============deployment ending====================

**Comment by peterwywong** [comment]
I could submit the new test cases as a separate PR. At any rate, three changes are now under one PR. Let me know if such a separation is needed.

For the diskless and diskfull provisionings in a hierarchical cluster, dir1/dir2/dir3/foo.bar (755) is added under /install/postscripts, "foo.bar 1" and "foo.bar 0" are added to postscripts correspondingly.

After provisioning, xcat.log is parsed to look for strings of successful runs of foo.bar.

Finally, foo.bar and the postscripts attribute are cleaned up before the next test case.

In /var/log/xcat/xcat.log on CN, the following are found:

Wed Apr 15 15:19:08 EDT 2020 [info]: xcat.deployment.postbootscript: postbootscript start..: dir1/dir2/dir3/foo.bar
"The number is ONE"
Wed Apr 15 15:19:08 EDT 2020 [info]: xcat.deployment.postbootscript: postbootscript end...:dir1/dir2/dir3/foo.bar return with 0
Wed Apr 15 15:19:08 EDT 2020 [info]: xcat.deployment.postbootscript: postbootscript start..: dir1/dir2/dir3/foo.bar
"The number is NOT 1"
Wed Apr 15 15:19:08 EDT 2020 [info]: xcat.deployment.postbootscript: postbootscript end...:dir1/dir2/dir3/foo.bar return with 0

foo.bar executes correctly and returns 0 in both cases.

**Comment by cxhong** [comment]
@peterwywong , I thought you are going to add a new test cases ,  @besawn , is right?

I don't think this test should be in the  `xCAT-test/autotest/testcase/installation/reg_linux_diskfull_installation_hierarchy` and `xCAT-test/autotest/testcase/installation/reg_linux_diskless_installation_hierarchy`

**Comment by peterwywong** [comment]
I was working on separate new test cases before Monday's xCAT meeting which you were unable to join.

I planned to add them at the end of the bundles.

Nate suggested that I embed my tests in some existing postscript test cases. However, these new tests must work around OS provisioning, so reg_linux_diskless_installation_hierarchy and reg_linux_diskfull_install_hierarchy were considered. Another consideration was to avoid additional OS provisioning times. It takes 35-40 minutes at the end of each bundle to run two new tests: one for diskfull and one for diskless.

Mark and I are in agreement with it. Therefore, I made changes accordingly.

With the first approach, there are two "new" test case files. And "a bunch of bundle files" need to be modified to include them.

With the second approach, there are two "modified" test case files. There is no need to modify any bundle files.

**Comment by peterwywong** [comment]
Regarding your change to the title:
The problem is not the execute permission to all files recursively under /install/postscripts on MN.
The problem is the execute permission to all files recursively under /xcatpost on CN.

**Comment by peterwywong** [comment]
Nate thinks that it is good to add this new test case updatenode_postscripts_permission to test "updatenode -P", along with the other two added yesterday.
By doing so, we need to include it in bundles.
Which bundles should I modify?

**Comment by peterwywong** [comment]
If we want to test it on diskless and diskfull, I can add it after reg_linux_diskless_installation_hierarchy and another one after reg_linux_diskfull_installation_hierarchy.
If testing once is sufficient, I can add it at the end of the bundle.

**Comment by cxhong** [comment]
ok, I see.  I don't have better suggestion.   Mark worked on the new test cases, I will sign him for review.

**Comment by peterwywong** [comment]
Casandra, my initial test case was based on confignetwork as suggested by you.
Now, they are part of reg_linux_disk[less|full]_installation_hierarchy.
As far as invoking retry_install is concerned, they are all similar.

File: xCAT-test/autotest/testcase/installation/reg_linux_diskfull_installation_hierarchy [hunk 1]
```diff
@@ -18,6 +18,15 @@ check:rc==0
 cmd:chtab key=nameservers site.value="<xcatmaster>"
 check:rc==0
 
+cmd:mkdir -p /install/postscripts/dir1/dir2/dir3
+check:rc==0
+cmd:echo -e '#!/bin/bash\n\nif [[ $1 == 1 ]] \nthen\n  echo \"The number is ONE\";\nelse\n  echo \"The number is NOT 1\"; \nfi\nexit 0;'> /install/postscripts/dir1/dir2/dir3/foo.bar
+check:rc==0
+cmd:chmod 755 /install/postscripts/dir1/dir2/dir3/foo.bar
+check:rc==0
+cmd:chdef -p -t node -o $$CN postscripts="dir1/dir2/dir3/foo.bar 1,dir1/dir2/dir3/foo.bar 0"
```
Comment (gurevichmark) [inline xCAT-test/autotest/testcase/installation/reg_linux_diskfull_installation_hierarchy:27]: What is the reason for running foo.bar twice, one with input 1 and then with input of 0 ?
It seems in both cases rc=0 will be returned and execute permission on the file would apply to both invocations ?

**Comment by peterwywong** [comment]
Mark, it is a good question.
Originally, we want to return 0 for input 1 and 1 for input 0.
For "updatenode -P", CHECK: rc==1 is tested.
However, right now, it returns 0 in both cases to avoid failure for retry_install,, so I think one can be removed since it loses its original purpose and "foo.bar 1" is enough to test the execute permission. 
Does it sound reasonable? If so, I modify them.

**Comment by gurevichmark** [comment]
@peterwywong Yes. I would even simplify it so it does not take any input.
Just print something with `echo` and return 0.

**Comment by peterwywong** [comment]
We still need to decide the bundle changes to include updatenode_postscripts_permission.

**Comment by peterwywong** [comment]
Let's hold the thought of changing bundles at this time. I now think there may be existing updatenode test cases which can do the permission testing too.

**Comment by peterwywong** [comment]
I now recognize that "updatenode_P_script1" in updatenode/cases0 can test whether the execute permission preserves through updatenode. And this test is probably included in all of the bundles.
If you agree on it, please let me know how can I roll back my updatenode changes?

As I work on the test cases more, I gradually see better :) .

**Comment by gurevichmark** [comment]
@peterwywong Modifying existing testcase `updatenode_P_script1` should work.

To undo the file changes to the file, you can try:
`git checkout -- <file>` - this will wipeout that file changes in you branch is replace the file with the latest committed version
`git add <file>`
`git commit ...`
`git push ...`

**Pull request merged**